test/e2e/storage/framework/volume_group_snapshot_resource.go

```diff
@@ -0,0 +1,126 @@
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+func getVolumeGroupSnapshot(groupName string, ns, snapshotClassName string) *unstructured.Unstructured {
+	snapshot := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       "VolumeGroupSnapshot",
+			"apiVersion": utils.VolumeGroupSnapshotAPIVersion,
+			"metadata": map[string]interface{}{
+				"generateName": "group-snapshot-",
+				"namespace":    ns,
+			},
+			"spec": map[string]interface{}{
+				"volumeGroupSnapshotClassName": snapshotClassName,
+				"source": map[string]interface{}{
+					"selector": map[string]interface{}{
+						"matchLabels": map[string]interface{}{
+							"group": groupName,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return snapshot
+}
+
+// VolumeGroupSnapshotResource represents a volumegroupsnapshot class, a volumegroupsnapshot and its bound contents for a specific test case
+type VolumeGroupSnapshotResource struct {
+	Config  *PerTestConfig
+	Pattern TestPattern
+
+	Vgs        *unstructured.Unstructured
+	Vgscontent *unstructured.Unstructured
+	Vgsclass   *unstructured.Unstructured
+}
+
+// CreateVolumeGroupSnapshot creates a VolumeGroupSnapshotClass with given SnapshotDeletionPolicy and a VolumeGroupSnapshot
+// from the VolumeGroupSnapshotClass using a dynamic client.
+// Returns the unstructured VolumeGroupSnapshotClass and VolumeGroupSnapshot objects.
+func CreateVolumeGroupSnapshot(ctx context.Context, sDriver VoulmeGroupSnapshottableTestDriver, config *PerTestConfig, pattern TestPattern, groupName string, pvcNamespace string, timeouts *framework.TimeoutContext, parameters map[string]string) (*unstructured.Unstructured, *unstructured.Unstructured) {
+	defer ginkgo.GinkgoRecover()
+	var err error
+	if pattern.SnapshotType != VolumeGroupSnapshot {
+		err = fmt.Errorf("SnapshotType must be set to VolumeGroupSnapshot")
+		framework.ExpectNoError(err)
+	}
+	dc := config.Framework.DynamicClient
+
+	ginkgo.By("creating a VolumeGroupSnapshotClass")
+	gsclass := sDriver.GetVolumeGroupSnapshotClass(ctx, config, parameters)
+	if gsclass == nil {
+		framework.Failf("Failed to get volume group snapshot class based on test config")
+	}
+	gsclass.Object["deletionPolicy"] = pattern.SnapshotDeletionPolicy.String()
+
+	gsclass, err = dc.Resource(utils.VolumeGroupSnapshotClassGVR).Create(ctx, gsclass, metav1.CreateOptions{})
+	framework.ExpectNoError(err)
+	gsclass, err = dc.Resource(utils.VolumeGroupSnapshotClassGVR).Get(ctx, gsclass.GetName(), metav1.GetOptions{})
+	framework.ExpectNoError(err)
+
+	ginkgo.By("creating a dynamic VolumeGroupSnapshot")
+	// Prepare a dynamically provisioned group volume snapshot with certain data
+	volumeGroupSnapshot := getVolumeGroupSnapshot(groupName, pvcNamespace, gsclass.GetName())
+
+	volumeGroupSnapshot, err = dc.Resource(utils.VolumeGroupSnapshotGVR).Namespace(volumeGroupSnapshot.GetNamespace()).Create(ctx, volumeGroupSnapshot, metav1.CreateOptions{})
+	framework.ExpectNoError(err)
```

Comment:
Can you please an add explanation for all `framework.ExpectNoError` ?

Reply:
Done